Mathlib/Tactic/FieldSimp.lean

```diff
@@ -0,0 +1,164 @@
+/-
+Copyright (c) 2019 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel, David Renshaw
+-/
+
+import Lean.Elab.Tactic.Basic
+import Lean.Meta.Tactic.Simp.Main
+import Std.Lean.Parser
+import Mathlib.Algebra.Group.Units
+import Mathlib.Tactic.NormNum.Core
+import Qq.Match
+
+/-!
+# `field_simp` tactic
+
+Tactic to clear denominators in algebraic expressions, based on `simp` with a specific simpset.
+-/
+
+namespace Mathlib.Tactic.FieldSimp
+
+open Lean Elab.Tactic Parser.Tactic Lean.Meta
+open Qq
+
+initialize registerTraceClass `Tactic.field_simp
+
+/-- Constructs a trace message for the `discharge` function. -/
+private def dischargerTraceMessage (prop: Expr) : Except ε (Option Expr) → SimpM MessageData
+| .error _ | .ok none => return m!"{crossEmoji} discharge {prop}"
+| .ok (some _) => return m!"{checkEmoji} discharge {prop}"
+
+/-- Discharge strategy for the field_simp tactic. -/
+partial def discharge (prop : Expr) : SimpM (Option Expr) :=
+  withTraceNode `Tactic.field_simp (dischargerTraceMessage prop) do
+    if let some r ← Simp.dischargeUsingAssumption? prop then
+      return some r
+
+    let prop : Q(Prop) ← (do pure prop)
+    let pf? ← match prop with
+    | ~q(($e : $α) ≠ $b) =>
```

Comment:
The mathlib3 version instead matches on `$e ≠ 0`. I couldn't figure out a way to make that work with `Qq`. (I could get it to typecheck here, but then it failed to actually match the expressions that we need it to. This more general match should be fine. The cost is that we might call `NormNum.derive` slightly more often than is strictly necessary.